src/libawkward/Content.cpp

```diff
@@ -1189,81 +1189,48 @@ namespace awkward {
-               dynamic_cast<IndexedOptionArray64*>(tmp1.get())) {
-        tmp2 = rawtmp1->project();
-        if (!check_missing_jagged_same(that, rawtmp1->bytemask(), missing)) {
-          return that;
-        }
-      }
-      else if (ByteMaskedArray* rawtmp1 =
-               dynamic_cast<ByteMaskedArray*>(tmp1.get())) {
-        tmp2 = rawtmp1->project();
-        if (!check_missing_jagged_same(that, rawtmp1->bytemask(), missing)) {
-          return that;
-        }
-      }
-      else if (BitMaskedArray* rawtmp1 =
-               dynamic_cast<BitMaskedArray*>(tmp1.get())) {
-        tmp2 = rawtmp1->project();
-        if (!check_missing_jagged_same(that, rawtmp1->bytemask(), missing)) {
-          return that;
-        }
-      }
-
-      if (tmp2.get() != nullptr) {
-        return std::make_shared<RegularArray>(Identities::none(),
-                                              that.get()->parameters(),
-                                              tmp2,
-                                              tmp2.get()->length());
-      }
+  const ContentPtr getitem_next_missing_jagged(const SliceMissing64& missing,
+                                               const Slice& tail,
+                                               const Index64& advanced,
+                                               const ContentPtr& that) {
+    const SliceJagged64* jagged = dynamic_cast<SliceJagged64*>(missing.content().get());
+    if ( jagged == nullptr ) {
+      throw std::runtime_error("Logic error: calling getitem_next_missing_jagged with bad slice type");
+    }
+    const Index64 index = missing.index();
+    ContentPtr content = that.get()->getitem_at_nowrap(0);
+    if ( content.get()->length() != index.length() ) {
+      throw std::invalid_argument(
+        std::string("cannot fit masked jagged slice with length ")
+        + std::to_string(index.length()) + std::string(" into ")
+        + that.get()->classname() + std::string(" of size ") + std::to_string(content.get()->length()));
     }
-    return that;
+    Index64 outputmask(index.length());
+    Index64 starts(index.length());
+    Index64 stops(index.length());
+    Content_getitem_next_missing_jagged_getmaskstartstop(index.ptr().get() + index.offset(),
+                                                         jagged->offsets().ptr().get() + jagged->offsets().offset(),
+                                                         outputmask.ptr().get(),
+                                                         starts.ptr().get(),
+                                                         stops.ptr().get(),
+                                                         index.length());
+    ContentPtr tmp = content.get()->getitem_next_jagged(starts, stops, jagged->content(), tail);
+    IndexedOptionArray64 out(Identities::none(), util::Parameters(), outputmask, tmp);
+    return std::make_shared<RegularArray>(Identities::none(), util::Parameters(), out.simplify_optiontype(), index.length());
```

Comment:
I cribbed this line from `getitem_next_regular_missing` but I can't say I understand why I have to re-wrap the array with a RegularArray, and why it has length `index.length()` rather than the length `1` that it came in with.

Reply:
I'm not close enough to the details at the moment to give advice. Is `index.length()` equal to `1` in all your examples? It might be equal all the time and the use of `index.length()` is just a matter of being formal.

Reply:
In fact index.length() is long, as long as the slice. It seems to be used also in https://github.com/scikit-hep/awkward-1.0/blob/master/src/libawkward/array/ListArray.cpp#L1644-L1646 so I think it is consistent with other uses. I guess a calling function unwraps it.